### PR TITLE
hermitian matrix multiplication (left lower)

### LINKED ISSUE #760: add hermitian matrix multiplication
Needed for result checking of eigensolver miniapps (#761).

### PR DESCRIPTION
**Pull request by rasolca**
Closes #760.

### CODE REVIEW

**Comment by rasolca** [comment]
cscs-ci run

File: include/dlaf/multiplication/hermitian/api.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include "dlaf/matrix/matrix.h"
+#include "dlaf/types.h"
+
+namespace dlaf {
+namespace multiplication {
+namespace internal {
```
Comment (msimberg) [inline include/dlaf/multiplication/hermitian/api.h:17]:
```suggestion
namespace dlaf::multiplication::internal {
```
?

File: include/dlaf/multiplication/hermitian/api.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+#ifdef DLAF_WITH_GPU
+DLAF_MULTIPLICATION_HERMITIAN_ETI(extern, Backend::GPU, Device::GPU, float)
+DLAF_MULTIPLICATION_HERMITIAN_ETI(extern, Backend::GPU, Device::GPU, double)
+DLAF_MULTIPLICATION_HERMITIAN_ETI(extern, Backend::GPU, Device::GPU, std::complex<float>)
+DLAF_MULTIPLICATION_HERMITIAN_ETI(extern, Backend::GPU, Device::GPU, std::complex<double>)
+#endif
+}
+}
+}
```
Comment (msimberg) [inline include/dlaf/multiplication/hermitian/api.h:44]: and
```suggestion
}
```
to match the first comment.

File: include/dlaf/multiplication/hermitian/impl.h [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <pika/execution.hpp>
+#include <pika/future.hpp>
```
Comment (msimberg) [inline include/dlaf/multiplication/hermitian/impl.h:13]:
```suggestion
#include <pika/future.hpp>
```
Not needed I think?

File: src/multiplication/hermitian/gpu.cpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include "dlaf/multiplication/hermitian/impl.h"
+
+namespace dlaf {
+namespace multiplication {
+namespace internal {
```
Comment (msimberg) [inline src/multiplication/hermitian/gpu.cpp:15]: Same as in the other files, but this is obviously not a blocker.

File: src/multiplication/hermitian/mc.cpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include "dlaf/multiplication/hermitian/impl.h"
+
+namespace dlaf {
+namespace multiplication {
+namespace internal {
```
Comment (msimberg) [inline src/multiplication/hermitian/mc.cpp:15]: And here.

**Review by msimberg — APPROVED** [review]
Looks good to me.

The namespace change is not a blocker. The `pika/future.hpp` header would be good to remove if it's indeed unused as it seems to be.

File: include/dlaf/multiplication/hermitian.h [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+#include "dlaf/types.h"
+#include "dlaf/util_matrix.h"
+
+namespace dlaf::multiplication {
+
+/// Hermitian Matrix multiplication implementation on local memory, computing C = beta C + alpha A B
+/// (when side == Left) or C + alpha B A (when side == Right), where A is a Hermitian matrix.
+///
+/// @param side specifies whether A appears on the \a Left or on the \a Right of matrix B,
+/// @param uplo specifies if the elements of the Hermitian matrix to be referenced are the elements in
```
Comment (albestro) [inline include/dlaf/multiplication/hermitian.h:25]: in ... now I'm curious 🤣

File: include/dlaf/multiplication/hermitian.h [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+/// @param mat_a contains the hermitian matrix A. Only the tiles of the matrix which contain the upper or
+/// the lower triangular part which represent the Hermitian matrix (depending on the value of uplo)
+/// are accessed in read-only mode (the elements are not modified),
+/// @param mat_b contains the matrix B accessed in read-only mode (the elements are not modified),
+/// @param mat_ic on entry it contains the matrix C, on exit the matrix elements are overwritten with the
```
Comment (albestro) [inline include/dlaf/multiplication/hermitian.h:30]:
```suggestion
/// @param mat_c on entry it contains the matrix C, on exit the matrix elements are overwritten with the
```

File: include/dlaf/multiplication/hermitian.h [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+/// @pre mat_a has a square block size,
+/// @pre mat_a mat_b and mat_c are not distributed,
+/// @pre mat_a mat_b are multipliable and the result can be summed to mat_c.
+template <Backend B, Device D, class T>
+void hermitian(blas::Side side, blas::Uplo uplo, const T alpha, Matrix<const T, D>& mat_a,
+               Matrix<const T, D>& mat_b, const T beta, Matrix<T, D>& mat_c) {
+  DLAF_ASSERT(matrix::square_size(mat_a), mat_a);
+  DLAF_ASSERT(matrix::square_blocksize(mat_a), mat_a);
+  DLAF_ASSERT(matrix::local_matrix(mat_a), mat_a);
+  DLAF_ASSERT(matrix::local_matrix(mat_b), mat_b);
```
Comment (albestro) [inline include/dlaf/multiplication/hermitian.h:42]:
```suggestion
  DLAF_ASSERT(matrix::local_matrix(mat_b), mat_b);
  DLAF_ASSERT(matrix::local_matrix(mat_c), mat_c);
```

**Review by albestro — APPROVED** [review]
LGTM, just a couple of comments.

**Comment by rasolca** [comment]
cscs-ci run

**Pull request merged**